# DOWNLOAD DATASET

In [81]:
!wget -O loan_train.csv https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/ML0101ENv3/labs/loan_train.csv

--2020-10-28 16:04:16--  https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/ML0101ENv3/labs/loan_train.csv
Resolving s3-api.us-geo.objectstorage.softlayer.net (s3-api.us-geo.objectstorage.softlayer.net)... 67.228.254.196
Connecting to s3-api.us-geo.objectstorage.softlayer.net (s3-api.us-geo.objectstorage.softlayer.net)|67.228.254.196|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 23101 (23K) [text/csv]
Saving to: ‘loan_train.csv’

loan_train.csv      100%[===================>]  22.56K  --.-KB/s    in 0.002s  

2020-10-28 16:04:16 (13.8 MB/s) - ‘loan_train.csv’ saved [23101/23101]



In [82]:
!wget -O loan_test.csv https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/ML0101ENv3/labs/loan_test.csv

--2020-10-28 16:04:17--  https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/ML0101ENv3/labs/loan_test.csv
Resolving s3-api.us-geo.objectstorage.softlayer.net (s3-api.us-geo.objectstorage.softlayer.net)... 67.228.254.196
Connecting to s3-api.us-geo.objectstorage.softlayer.net (s3-api.us-geo.objectstorage.softlayer.net)|67.228.254.196|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 3642 (3.6K) [text/csv]
Saving to: ‘loan_test.csv’

loan_test.csv       100%[===================>]   3.56K  --.-KB/s    in 0s      

2020-10-28 16:04:17 (88.4 MB/s) - ‘loan_test.csv’ saved [3642/3642]



# IMPORT LIBRARIES

In [83]:
import pandas as pd
from sklearn.preprocessing import LabelEncoder
import seaborn as sns
from sklearn.metrics import jaccard_score
from sklearn.metrics import f1_score
from sklearn.metrics import log_loss
import itertools
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.ticker import NullFormatter
import pandas as pd
import numpy as np
import matplotlib.ticker as ticker
from sklearn import preprocessing
%matplotlib inline

# IMPORT DATA INTO PANDAS DATAFRAME

In [84]:
#Training data set
df_train = pd.read_csv('loan_train.csv')
df_train.head()

Unnamed: 0  Unnamed: 0.1 loan_status  Principal  terms effective_date  \
0           0             0     PAIDOFF       1000     30       9/8/2016   
1           2             2     PAIDOFF       1000     30       9/8/2016   
2           3             3     PAIDOFF       1000     15       9/8/2016   
3           4             4     PAIDOFF       1000     30       9/9/2016   
4           6             6     PAIDOFF       1000     30       9/9/2016   

    due_date  age             education  Gender  
0  10/7/2016   45  High School or Below    male  
1  10/7/2016   33              Bechalor  female  
2  9/22/2016   27               college    male  
3  10/8/2016   28               college  female  
4  10/8/2016   29               college    male

In [85]:
# Test data set
df_test = pd.read_csv('loan_test.csv')
df_test.head()

Unnamed: 0  Unnamed: 0.1 loan_status  Principal  terms effective_date  \
0           1             1     PAIDOFF       1000     30       9/8/2016   
1           5             5     PAIDOFF        300      7       9/9/2016   
2          21            21     PAIDOFF       1000     30      9/10/2016   
3          24            24     PAIDOFF       1000     30      9/10/2016   
4          35            35     PAIDOFF        800     15      9/11/2016   

    due_date  age             education  Gender  
0  10/7/2016   50              Bechalor  female  
1  9/15/2016   35       Master or Above    male  
2  10/9/2016   43  High School or Below  female  
3  10/9/2016   26               college    male  
4  9/25/2016   29              Bechalor    male

# VIEWING CATEGORIES FOR CATEGORICAL VARIABLES

In [86]:
df_train['loan_status']=pd.Categorical(df_train['loan_status'])
print(df_train['loan_status'].cat.categories)

Index(['COLLECTION', 'PAIDOFF'], dtype='object')


In [87]:
df_train['education']=pd.Categorical(df_train['education'])
print(df_train['education'].cat.categories)

Index(['Bechalor', 'High School or Below', 'Master or Above', 'college'], dtype='object')


In [88]:
df_train['Gender']=pd.Categorical(df_train['Gender'])
print(df_train['Gender'].cat.categories)

Index(['female', 'male'], dtype='object')


In [89]:
df_test['loan_status']=pd.Categorical(df_test['loan_status'])
print(df_test['loan_status'].cat.categories)

Index(['COLLECTION', 'PAIDOFF'], dtype='object')


In [90]:
df_test['loan_status']=pd.Categorical(df_test['loan_status'])
print(df_test['loan_status'].cat.categories)

Index(['COLLECTION', 'PAIDOFF'], dtype='object')


In [91]:
df_test['loan_status']=pd.Categorical(df_test['loan_status'])
print(df_test['loan_status'].cat.categories)

Index(['COLLECTION', 'PAIDOFF'], dtype='object')


# DATA CLEANING

# CONVERING CATEGORICAL VARIBLES TO NUMERICAL VARIABLES

In [92]:
df_train['Gender'].replace(to_replace=['male','female'], value=[0,1],inplace=True)

In [93]:
df_test['Gender'].replace(to_replace=['male','female'], value=[0,1],inplace=True)

In [94]:
df_train['loan_status'].replace(to_replace=['PAIDOFF','COLLECTION'], value=[0,1],inplace=True)

In [95]:
df_test['loan_status'].replace(to_replace=['PAIDOFF','COLLECTION'], value=[0,1],inplace=True)

# CONVERTING ORDINAL VARIABLES TO NUMERICAL

In [96]:
Feature_train = df_train[['Principal','terms','age','Gender']]
Feature_train = pd.concat([Feature_train,pd.get_dummies(df_train['education'])], axis=1)
Feature_train.drop(['Master or Above'], axis = 1,inplace=True)
Feature_train.head()

Principal  terms  age  Gender  Bechalor  High School or Below  college
0       1000     30   45       0         0                     1        0
1       1000     30   33       1         1                     0        0
2       1000     15   27       0         0                     0        1
3       1000     30   28       1         0                     0        1
4       1000     30   29       0         0                     0        1

In [97]:
Feature_test = df_test[['Principal','terms','age','Gender']]
Feature_test = pd.concat([Feature_test,pd.get_dummies(df_test['education'])], axis=1)
Feature_test.drop(['Master or Above'], axis = 1,inplace=True)
Feature_test.head()

Principal  terms  age  Gender  Bechalor  High School or Below  college
0       1000     30   50       1         1                     0        0
1        300      7   35       0         0                     0        0
2       1000     30   43       1         0                     1        0
3       1000     30   26       0         0                     0        1
4        800     15   29       0         1                     0        0

# FEATURE SELECTION

In [98]:
x_train = Feature_train

In [99]:
from sklearn import preprocessing
x_test = Feature_test

In [100]:
y_train = df_train['loan_status'].values

In [101]:
y_test = df_test['loan_status'].values

# APPLYING THE MACHINE LEARNING MODELS

# k-Nearest Neighbour

In [102]:
from sklearn.neighbors import KNeighborsClassifier
k = 3
kNN_model = KNeighborsClassifier(n_neighbors=k).fit(x_train,y_train)
knn_yhat = kNN_model.predict(x_test)
print("KNN Jaccard index: %.2f" % jaccard_score(y_test, knn_yhat))
print("KNN F1-score: %.2f" % f1_score(y_test, knn_yhat, average='weighted') )

KNN Jaccard index: 0.07
KNN F1-score: 0.66


# Decision Tree

In [103]:
from sklearn.tree import DecisionTreeClassifier
DT_model = DecisionTreeClassifier(criterion="entropy", max_depth = 4)
DT_model.fit(x_train,y_train)
DT_yhat = DT_model.predict(x_test)
print("DT Jaccard index: %.2f" % jaccard_score(y_test, DT_yhat))
print("DT F1-score: %.2f" % f1_score(y_test, DT_yhat, average='weighted') )

DT Jaccard index: 0.00
DT F1-score: 0.63


# Support Vector Machine

In [104]:
from sklearn import svm
SVM_model = svm.SVC()
SVM_model.fit(x_train, y_train)
SVM_yhat = SVM_model.predict(x_test)
print("SVM Jaccard index: %.2f" % jaccard_score(y_test, SVM_yhat))
print("SVM F1-score: %.2f" % f1_score(y_test, SVM_yhat, average='weighted') )

SVM Jaccard index: 0.00
SVM F1-score: 0.63


# LOGISTIC REGRESSION

In [105]:
from sklearn.linear_model import LogisticRegression
LG_model = LogisticRegression(random_state=0).fit(x_train, y_train)
LG_yhat = LG_model.predict(x_test)
LG_yhat_prob = LG_model.predict_proba(x_test)
print("LG Jaccard index: %.2f" % jaccard_score(y_test, LG_yhat))
print("LG F1-score: %.2f" % f1_score(y_test, LG_yhat, average='weighted') )
print("LR LogLoss: %.2f" % log_loss(y_test, LG_yhat_prob))

LG Jaccard index: 0.00
LG F1-score: 0.63
LR LogLoss: 0.58


KNN Jaccard index: 0.07
KNN F1-score: 0.66
    
    
DT Jaccard index: 0.00
DT F1-score: 0.63
    
SVM Jaccard index: 0.00
SVM F1-score: 0.63
    
LG Jaccard index: 0.00
LG F1-score: 0.63
LR LogLoss: 0.58